### PR TITLE
Calculate party public keys during signature aggregation

### PR DESCRIPTION
**Pull request by xoloki**
Currently, each party sends its public key(s) with the `SignatureShare`.  This is unnecessary, since the public keys can be computed during signature aggregation.  This also lets us determine if the public key is bad, separately from the party signatures.

This also allows us to remove the generic arg from `SignatureShare`, and hence the `Signer` trait.  Doing so will allow us to make a `v2::Signer` object, which will make moving from `v1` to `v2` much easier.

In place of the public key, `SignatureShare` now has a set of `key_ids`.  For `v1`, this will be a single `key_id` equal to the `party_id`.  For `v2`, this will be the set of `key_ids` which the `Party` controls.

This PR also adds the necessary funcs to `traits::Signer` so it can be used with `v1` and `v2`, updates the `bip340::test_helpers` to take a generic `Signer` parameter which implements `traits::Signer` and only use funcs available to `Signer`, and adds a `bip340` `v2` test.

### CODE REVIEW

File: src/bip340.rs [hunk 1]
```diff
@@ -122,28 +120,18 @@ pub mod test_helpers {
...
-                for (id, share) in &broadcast_shares {
-                    h.insert(*id, share[&party.id]);
-                }
-
-                if let Err(secret_error) = party.compute_secret(h, &A) {
-                    secret_errors.insert(party.id, secret_error);
+            if let Err(signer_secret_errors) = signer.compute_secrets(&private_shares, &A) {
+                for (signer_id, errors) in signer_secret_errors.iter() {
+                    secret_errors.insert(*signer_id, errors.clone());
                 }
```
Comment (netrome) [inline src/bip340.rs:135]:
```suggestion
                secret_errors.extend(signer_secret_errors.into_iter());
```

**Review by netrome — APPROVED** [review]
LGTM, just one nit suggestion

File: src/v1.rs [hunk 2]
```diff
@@ -323,40 +334,85 @@ impl Signer {
...
+        key_ids: &[usize],
+        nonces: &[PublicNonce],
+    ) -> (Vec<Point>, Point) {
+        compute::intermediate(msg, key_ids, nonces)
+    }
+
+    fn sign(
+        &self,
+        msg: &[u8],
+        _signer_ids: &[usize],
```
Comment (jferrant) [inline src/v1.rs:409]: Is _signer_ids passed for an upcoming change?

File: src/bip340.rs [hunk 3]
```diff
@@ -231,13 +220,70 @@ mod test {
...
+            .collect();
+
+        let A = match test_helpers::dkg(&mut signers, &mut rng) {
+            Ok(A) => A,
+            Err(secret_errors) => {
+                panic!("Got secret errors from DKG: {:?}", secret_errors);
+            }
+        };
+
+        let mut S = [signers[0].clone(), signers[1].clone(), signers[3].clone()].to_vec();
```
[the remark this answers is not in the document]
Comment (xoloki) [reply src/bip340.rs:263]: This is actually a discontinuous set of selected signers: `[0, 1, 3]` out of `[0..4]`, whereas your suggestion would give `[0..3]`

[the remark this answers is not in the document]
Comment (xoloki) [reply src/bip340.rs:263]: Actually, I think your suggestion gives `[0, 1, 2]`.  Either way, it's not `[0, 1, 3]`.  

Unless I'm missing something, no sleep XD

[the remark this answers is not in the document]
Comment (xoloki) [reply src/bip340.rs:261]: Oh wait I think I just made a distinction without a difference XD

**Pull request merged**